array.go

```diff
@@ -1197,6 +1197,143 @@ func (a *Array) NotContainsAny(values ...interface{}) *Array {
+// Built-in comparator requires all elements in the array to have same data type.
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{100, 101, 102})
+//	array.IsOrdered() // succeeds
+//	array.IsOrdered(func(x, y *httpexpect.Value) bool {
+//		return x.Number().Raw() < y.Number().Raw()
+//	}) // succeeds
+func (a *Array) IsOrdered(less ...func(x, y *Value) bool) *Array {
+	a.chain.enter("IsOrdered()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if len(less) > 1 {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected multiple less arguments"),
+			},
+		})
+		return a
+	}
+
+	if len(a.value) == 0 {
+		return a
+	}
+
+	var lessFn func(x, y *Value) bool
+	if len(less) == 1 {
+		lessFn = less[0]
+	} else {
+		lessFn = a.getBuiltinComparator()
+		if a.chain.failed() {
+			return a
+		}
+	}
+
+	for i := 0; i < int(len(a.value))-1; i++ {
+		xChain := a.chain.clone()
+		xChain.replace("IsOrdered[%d]", i)
+		x := newValue(xChain, a.value[i])
+
+		yChain := a.chain.clone()
+		yChain.replace("IsOrdered[%d]", i+1)
+		y := newValue(yChain, a.value[i+1])
+
+		if lessFn(y, x) {
+			a.chain.fail(AssertionFailure{
+				Type:      AssertLt,
+				Actual:    &AssertionValue{x.value},
+				Expected:  &AssertionValue{y.value},
+				Reference: &AssertionValue{a.value},
+				Errors: []error{
+					errors.New("expected: array is ordered"),
+					fmt.Errorf("element %v must be less than element %v, but it isn't",
+						x.value, y.value),
+				},
+			})
+			return a
+		}
+	}
+
+	return a
+}
+
+func (a *Array) getBuiltinComparator() func(x, y *Value) bool {
+	var fn func(x, y *Value) bool
+	var prev interface{}
+	for _, curr := range a.value {
+		switch curr.(type) {
+		case bool, float64, string:
+			// do nothing
+			// for better performance, function is generated after validating all elements
+		default:
+			a.chain.fail(AssertionFailure{
+				Type:      AssertBelongs,
+				Actual:    &AssertionValue{fmt.Sprintf("%T", curr)},
+				Expected:  &AssertionValue{AssertionList{"Boolean", "Number", "String"}},
+				Reference: &AssertionValue{curr},
+				Errors: []error{
+					errors.New("expected: type of each element of reference array" +
+						"belongs to given list"),
+					fmt.Errorf("element %v has type %s", curr, fmt.Sprintf("%T", curr)),
+				},
+			})
+			return fn
+		}
+		if prev != nil && fmt.Sprintf("%T", curr) != fmt.Sprintf("%T", prev) {
+			a.chain.fail(AssertionFailure{
+				Type:      AssertEqual,
+				Actual:    &AssertionValue{fmt.Sprintf("%T", curr)},
+				Expected:  &AssertionValue{fmt.Sprintf("%T", prev)},
+				Reference: &AssertionValue{curr},
+				Errors: []error{
+					errors.New("expected: types of all elements of reference array are the same"),
+					fmt.Errorf("previous element %v has type %s and current element %v has type %s",
+						prev, fmt.Sprintf("%T", prev), curr, fmt.Sprintf("%T", curr)),
```

Comment:
This can be simplified the same way as mentioned above.

Also, let's remove "previous" and "current" from error message, the user has no idea that there is some "current" element.